### PR TITLE
sql/row: fix updates of single-composite-column families

### LINKED ISSUE #131860: sql/row: not deleting single-column family of composite column when updating with non-composite value
This is pretty niche, but this `continue` should sometimes be an `if` branch over MarshalLegacy if the reason we're skipping is because a value is non-composite:

https://github.com/cockroachdb/cockroach/blob/78d326c48c2420760c59ad6bcf647670919b1e38/pkg/sql/row/writer.go#L142

Here's an example:

```sql
CREATE TABLE abc (a INT NOT NULL, b FLOAT NOT NULL, c INT, FAMILY (a), FAMILY (b), FAMILY (c));
-- note that -0::float is a composite value if b becomes part of the PK
INSERT INTO abc VALUES (4, -0, 6);
-- make b part of the PK
ALTER TABLE abc ADD PRIMARY KEY (a, b);
-- for example, this deletes the c column family KV as it should
UPDATE abc SET c = NULL WHERE a = 4 AND b = -0;
-- but this does not delete the b column family KV when b is no longer a composite value
UPDATE abc SET b = 0 WHERE a = 4 AND b = -0;
-- so now this still shows -0 (the composite value)
SELECT * FROM abc;
```

Looks like this was introduced in v22.2 from 5dff0e5baa1cdca1ea9f9eb317d6aed290920896

Jira issue: CRDB-42727

### PR DESCRIPTION
**Pull request by michae2**
When updating a single-column family which contains what could be a composite value from the primary key, we still need to issue a Del even if the new value for the column is not composite, because the previous value might have been composite.

Fixes: #131860
Informs: #131645

Release note (bug fix): Fix a rare bug in which an update of a primary key column which is also the only column in a separate column family can sometimes fail to update the primary index. This bug has existed since v22.2.

### CODE REVIEW

**Comment by michae2** [comment]
(I noticed this while reviewing https://github.com/cockroachdb/cockroach/pull/131645.)

File: pkg/sql/row/writer.go [hunk 1]
```diff
@@ -136,31 +136,43 @@ func prepareInsertOrUpdateBatch(
...
+					}
+					oldVal = old.TagAndDataBytes()
 				}
-				oldVal = old.TagAndDataBytes()
 			}
 
-			if marshaled.RawBytes == nil {
-				if overwrite {
+			if len(marshaled.RawBytes) == 0 {
+				if overwrite || oth.IsSet() {
```
Comment (stevendanna) [inline pkg/sql/row/writer.go:175]: I can always do this in my follow-up, but I wonder what you think about this, just so future us have a reminder of why we were doing this.  I think all of the family.ID != 0 paths are suspect at the moment as they aren't covered by our LDR tests since we disable multiple column families.

```golang
				if oth.IsSet() {
					// If using OriginTimestamp'd CPuts, we _always_ want to issue a Delete
					// so that we can confirm our expected bytes were correct.
					oth.DelWithCPut(ctx, batch, kvKey, oldVal, traceKV)
				} else if overwrite {
					// If the new family contains a NULL value, then we must
					// delete any pre-existing row.
					insertDelFn(ctx, batch, kvKey, traceKV)
				}
```

File: pkg/sql/row/writer.go [hunk 2]
```diff
@@ -136,31 +136,43 @@ func prepareInsertOrUpdateBatch(
...
 
 			// TODO(ssd): Here and below investigate reducing the
 			// number of allocations required to marshal the old
 			// value.
 			var oldVal []byte
 			if oth.IsSet() && len(oldValues) > 0 {
-				old, err := valueside.MarshalLegacy(typ, oldValues[idx])
-				if err != nil {
-					return nil, err
+				if !couldBeComposite || oldValues[idx].(tree.CompositeDatum).IsComposite() {
```
Comment (stevendanna) [inline pkg/sql/row/writer.go:165]: I think we can't go wrong with a comment here, because to be honest I'm finding it hard to follow. Let me try to think this through in realtime here in this comment:

We are encoding a single-column, non-family-0 column family.  So, our _expected bytes_ should be:

1. Nil if the value is NULL.
2. Nil is the value is non-NULL but exists _complete_ in the primary key (i.e. is in the primary key and is not a composite value).
3. Non nil otherwise.

Unless I'm reading wrongly, I think this code probably is not quite right for case 1. since I think `TagAndDataBytes` will explode if old was encoded with just nil, so we might have to wrap that in an 'IsPresent()' call.

**Review by stevendanna — COMMENTED** [review]
Great catch.  In the other PR I'm going to try to write some tests for these multi-column-family cases even though we can't hit them in production yet.

File: pkg/sql/row/writer.go [hunk 3]
```diff
@@ -136,31 +136,43 @@ func prepareInsertOrUpdateBatch(
...
+						return nil, err
+					}
+					oldVal = old.TagAndDataBytes()
 				}
-				oldVal = old.TagAndDataBytes()
 			}
 
-			if marshaled.RawBytes == nil {
-				if overwrite {
+			if len(marshaled.RawBytes) == 0 {
```
[the remark this answers is not in the document]
Comment (michae2) [reply pkg/sql/row/writer.go]: done

File: pkg/sql/row/writer.go (hunk 2 above)
Comment (michae2) [reply pkg/sql/row/writer.go:167]: Added a little comment, and wrapped the call to `TagAndDataBytes` in a check for `IsPresent`.

(I'll leave the assertions up to you in your PR.)

**Review by stevendanna — APPROVED** [review]

**Comment by michae2** [comment]
Not sure what order we want to merge, but I'll just go ahead...

bors r=stevendanna
